### PR TITLE
Handle timeout on UART for Redox Wireless receiver-to-keyboard communication.

### LINKED ISSUE #16553: redox_w firmware broke with switch to UART
I could no longer produce a working firmware via configurator, while it worked before. I've built maually and verified that it broke between 0.15.25 (works) and 0.16 (does not work), and head also doesn't work. I've tried git bisect but must not be doing it right yet. I'll keep searching to try to figure out the cause.

The bad firmware flashes successfully but no keypresses register.

### PR DESCRIPTION
**Pull request by tomaszj**
- This fixes the issue of keyboard deadlocking on the first matrix
  scan with Redox Wireless keyboards

## Description

Resolves #16553. Closes #17071 

This brings back handling time-out in matrix scanning code. Versions `0.15.x` and before handled timeout implicitly, by using a non-blocking `UDR` register read.

The timeout variable is shared across each matrix scan — happy to change it to be individual to each UART byte read instead on a clear recommendation from someone more knowledgeable in UART comms.

The current `HEAD` uses a blocking read, disregarding the timeout code.

From my testing, it looks like it's only ever the first write/receive loop that fails. Debug messages never occured later in the short space I've tested this.

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* #16553 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by akavel** [comment]
FWIW, I've been using the keyboard for the whole day with this PR, and it seems to fix the issue for me - thanks a lot! ❤️

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/redox_w/matrix.c [hunk 1]
```diff
@@ -39,11 +41,16 @@ bool matrix_scan_custom(matrix_row_t current_matrix[]) {
...
+            if (timeout > UART_MATRIX_RESPONSE_TIMEOUT) {
                 break;
             }
         }
-        uart_data[i] = uart_read();
+
+        if (timeout < UART_MATRIX_RESPONSE_TIMEOUT) {
+            uart_data[i] = uart_read();
+        } else {
+            uart_data[i] = (uint8_t) 0x00;
```
[the remark this answers is not in the document]
Comment (tomaszj) [reply keyboards/redox_w/matrix.c:52]: Added that to be explicit and trigger a quick chat about the convention for fixed values. Happy to remove :)

**Comment by mffseal** [comment]
Thanks, It's works for my redox_w keyboard.
I applied the pr in my local qmk msys environment and successfuly compiled the available firmware(use JSON file generated by qmk configurator then use json2c to convert it to keymap.c).

**Comment by tomaszj** [comment]
@fauxpark The change is isolated to `redox_w` and in current form doesn't impact `mitosis`, etc. Once merged, I'm happy to prep a collective PR for these boards to keep code consistent and test with owners of said boards.

**Comment by fauxpark** [comment]
Yes, that's what I meant.

**Comment by drashna** [comment]
Thanks!

**Pull request merged**